### PR TITLE
feat: add interpolation fill strategy to GapFillExec

### LINKED ISSUE #6975: Add `interpolate` capability to `GapFillExec` without `prev` or `next`
Part of #6594. 

# Implementation Notes

`Interpolate` will be another fill strategy that will live alongside of `Null`, `Prev`, and `PrevNullAsMissing`:
https://github.com/influxdata/influxdb_iox/blob/866f9cefa1386f2c4ef1a07aadecb074f6c7835d/iox_query/src/exec/gapfill/mod.rs#L63-L76

This issue will treat nulls in the input as breaking and will not interpolate between two non-null points if there is a null in between. This is not quite the InfluxQL behavior. For that, see #7287.

### PR DESCRIPTION
**Pull request by wolffcm**
Closes #6975.

This implements a new fill strategy for `GapFillExec`. It is implemented in a similar fashion to the `Prev` fill strategy. I put it in a separate module since `algo.rs` was getting pretty big.

### CODE REVIEW

File: iox_query/src/exec/gapfill/algo/interpolate.rs [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+#[derive(Clone, Debug)]
+pub struct Segment<T> {
+    start_point: (i64, T),
+    end_point: (i64, T),
+}
+
+/// A macro to go from `Segment<$NATIVE>` into [`Segment<ScalarValue>`].
+/// Between output batches data values in segments are stored as [`ScalarValue`]
+/// to avoid type parameters in [`Cursor`].
+macro_rules! impl_try_from_segment_native {
```
Comment (wolffcm) [inline iox_query/src/exec/gapfill/algo/interpolate.rs:128]: I found that I had to resort to a macro here since any `impl From` or `impl TryFrom` that had a type parameter would conflict with the blanket implementation in the Rust stdlib.

File: iox_query/src/exec/gapfill/algo/interpolate.rs [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+    /// Create an Arrow array with gaps filled in between values
+    /// using linear interpolation.
+    pub(super) fn build_aggr_fill_interpolate(
+        &mut self,
+        params: &GapFillParams,
+        series_ends: &[usize],
+        input_time_array: &TimestampNanosecondArray,
+        input_aggr_array: &ArrayRef,
+    ) -> Result<ArrayRef> {
+        match input_aggr_array.data_type() {
```
Comment (alamb) [inline iox_query/src/exec/gapfill/algo/interpolate.rs:32]: You might be able to use https://docs.rs/arrow-array/35.0.0/arrow_array/macro.downcast_primitive_array.html to make this less verbose.

File: iox_query/src/exec/gapfill/algo/interpolate.rs [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+            DataType::Int64 => {
+                let input_aggr_array = input_aggr_array
+                    .as_any()
+                    .downcast_ref::<Int64Array>()
+                    .unwrap();
```
Comment (alamb) [inline iox_query/src/exec/gapfill/algo/interpolate.rs:37]: As an FYI you can use https://docs.rs/arrow-array/35.0.0/arrow_array/cast/fn.as_primitive_array.html here too if you wanted slightly more concision

```suggestion
                let input_aggr_array = as_primitive_array<Int64Type>
                  (input_aggr_array);

```

File: iox_query/src/exec/gapfill/algo/interpolate.rs (hunk 1 above)
Comment (alamb) [reply iox_query/src/exec/gapfill/algo/interpolate.rs:128]: Maybe @tustvold  has some clever ideas on how to (ab)use the Rust type system

File: iox_query/src/exec/gapfill/algo/interpolate.rs [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+        algo::tests::{array_to_lines, assert_cursor_end_state, new_cursor_with_batch_size},
+        params::GapFillParams,
+        FillStrategy,
+    };
+
+    /// Verify the rounding behavior (really just truncating towards zero) which is
+    /// what InfluxQL does. Also verify that we can have a descending slope in the
+    /// line that does not overflow a `u64`.
+    #[test]
+    fn test_interolate_u64() {
```
Comment (alamb) [inline iox_query/src/exec/gapfill/algo/interpolate.rs:327]:
```suggestion
    fn test_interpolate_u64() {
```

File: iox_query/src/exec/gapfill/algo/interpolate.rs [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+        - "| 1970-01-01T00:00:00.000001900Z | 0    |"
+        - "| 1970-01-01T00:00:00.000002Z    |      |"
+        - +--------------------------------+------+
+        "###);
+
+        assert_cursor_end_state(&cursor, &input_times, &params);
+    }
+
+    #[test]
+    fn test_interolate_i64() {
```
Comment (alamb) [inline iox_query/src/exec/gapfill/algo/interpolate.rs:399]:
```suggestion
    fn test_interpolate_i64() {
```

File: iox_query/src/exec/gapfill/algo/interpolate.rs [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+        - "| time                           | a0   |"
+        - +--------------------------------+------+
+        - "| 1970-01-01T00:00:00.000001Z    |      |"
+        - "| 1970-01-01T00:00:00.000001100Z | 100  |"
+        - "| 1970-01-01T00:00:00.000001200Z | 133  |"
+        - "| 1970-01-01T00:00:00.000001300Z | 166  |"
+        - "| 1970-01-01T00:00:00.000001400Z | 200  |"
+        - "| 1970-01-01T00:00:00.000001500Z |      |"
+        - "| 1970-01-01T00:00:00.000001600Z |      |"
+        - "| 1970-01-01T00:00:00.000001700Z | 1000 |"
```
Comment (alamb) [inline iox_query/src/exec/gapfill/algo/interpolate.rs:388]: I expected these two rows to be interpolated between 200 and 1000 -- so making up `800` over 3 slots they should  have a values of `466` and `732`, right?

The same question applies to the other tests too

File: iox_query/src/exec/gapfill/exec_tests.rs [hunk 7]
```diff
@@ -752,6 +752,116 @@ fn test_gapfill_fill_prev_null_as_missing_many_nulls() {
...
+                - +----+--------------------------+------+
+                - "| g0 | time                     | a0   |"
+                - +----+--------------------------+------+
+                - "| a  |                          | -1   |"
+                - "| a  | 1970-01-01T00:00:00.975Z |      |"
+                - "| a  | 1970-01-01T00:00:01Z     | 100  |"
+                - "| a  | 1970-01-01T00:00:01.025Z | 200  |"
+                - "| a  | 1970-01-01T00:00:01.050Z | 300  |"
+                - "| a  | 1970-01-01T00:00:01.075Z | 400  |"
+                - "| a  | 1970-01-01T00:00:01.100Z |      |"
```
Comment (alamb) [inline iox_query/src/exec/gapfill/exec_tests.rs:847]: I don't understand why these are null -- is it because there are no more non-null observations in the group?

File: iox_query/src/exec/gapfill/mod.rs [hunk 8]
```diff
@@ -73,6 +73,11 @@ pub enum FillStrategy {
     /// This is the InfluxQL behavior for `FILL(PREVIOUS)`.
     #[allow(dead_code)]
     PrevNullAsMissing,
+    /// Fill the gaps between points linearly.
+    /// Null values will not be considered as missing, so two non-null values
+    /// with a null in between will not be filled.
+    #[allow(dead_code)]
```
Comment (alamb) [inline iox_query/src/exec/gapfill/mod.rs:79]: Is the `allow(dead_code)` still needed?

**Review by alamb — COMMENTED** [review]
All in all this is great stuff @wolffcm  -- it looks great to me other than one question about the output of interpolating in the tests (I don't understand why some values are null).

Also, would it be possible to add a SQL level end to end test for this feature? Or maybe that needs some frontend work to support the `interpolate()` function

While reviewing this I couldn't help but remember some of the Window function implementation that is done in DataFusion and is somewhat similar to this code. The WindowExec track windows across an ordered list of rows, and and incrementally updates `WindowExpr`s as the values go in / out of the aggregator

https://docs.rs/datafusion/latest/datafusion/physical_plan/windows/struct.WindowAggExec.html

https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.WindowExpr.html

I was wondering if we could reuse `WindowExec` somehow 🤔 

Something like 
```text
                                                   ┌───────────────────────┐      
                                                   │      WindowExec       │      
                                                   └───────────────────────┘      
                                                               ▲                  
                                                               │                  
  ┌───────────────────────┐                        ┌───────────────────────┐      
  │      GapFillExec      │                        │        GapFill        │      
  └───────────────────────┘                        └───────────────────────┘      
                                                               ▲                  
                                                               │                  
  ┌───────────────────────┐                        ┌───────────────────────┐      
  │         Scan          │                        │         Scan          │      
  └───────────────────────┘                        └───────────────────────┘      
                                                                                  
Approach in this PR is do to the                Another approach might be to leave
interpolation in the GapFill node               the GapFill node to fill gaps, and
                                                then use WindowExec / WindowExpr  
                                                functions to fill gaps (nulls)    
                                                                  
```

The upside to using the WindowExec approach is that there was quite a lot of effort put into them to make them efficient and parallelized (partitioned) and streaming

At this point we probably shouldn't rip out all the interpolation logic, but if we want to add lots more interpolation logic (e.g. what @stuartcarnie is thinking about for InfluxQL) it might be worth a larger discussion

File: iox_query/src/exec/gapfill/algo/interpolate.rs (hunk 3 above)
Comment (wolffcm) [reply iox_query/src/exec/gapfill/algo/interpolate.rs:37]: I like concision, I will do it, thank you.

File: iox_query/src/exec/gapfill/algo/interpolate.rs (hunk 1 above)
Comment (wolffcm) [reply iox_query/src/exec/gapfill/algo/interpolate.rs:128]: After spending some time with it, my sense is (I would be glad to be wrong) that this is the best I could do. There is similar code in DataFusion for converting `ScalarValue` to and from native types.

**Comment by wolffcm** [comment]
@alamb 
> would it be possible to add a SQL level end to end test for this feature? Or maybe that needs some frontend work to support the interpolate() function

Not quite yet. For both filling with previous (`LOCF`) and (`INTERPOLATE`) the planner rule needs to be updated to find a projection with these functions and push them into the `GapFill` node. I have been trying to prioritize work that would be useful to InfluxQL, and it wasn't clear to me if @stuartcarnie's work would be making use of the planner rule or constructing the `GapFill` node directly.

It's looking like the planner rule will be used from the InfluxQL side as well though. My plan was to move on to updating the planner rule (https://github.com/influxdata/influxdb_iox/issues/7071) regardless, as it this functionality is really important for SQL as well.

> I was wondering if we could reuse `WindowExec` somehow

I really like this idea. I have been thinking about this since I read your comments yesterday. What you suggest might be possible, but as you might expect there is some subtlety there. I will provide some more context below.

---

There are really two ways to think about null values when gap-filling.

## Null as intentional

The user put the null values in the input on purpose, and doesn't want them filled. For TimeScale, this seems to be the default. For InfluxDB, since line protocol doesn't allow you create nulls except by omission, maybe this behavior is less useful at least for now.

In general, this treatment of nulls is easier to implement in `GapFillExec`, since it makes it easy to know how much of the input data we need to buffer to produce output.

However, this treatment of nulls would be harder to implement in a subsequent `WindowExec` node. How would it know which nulls were in the original input, and which need to be filled?

## Null as missing

The null values represent just another gap in the column and should be filled. TimeScale provides a `nullAsMissing` parameter to `LOCF` to get this behavior. This is also the behavior of `FILL(PREVIOUS)` and `FILL(LINEAR)` in InfluxQL.

This treatment of nulls is trickier to implement in `GapFillExec` because:
- We might need a strategy for holding onto older data (hence the "stash" mechanism from #7245)
- We might need to look forward an arbitrary amount to find the next non-null point (this would be the case with `INTERPOLATE` for #7287)

But if this would be easier to do in a `WindowExec` node, that could be a big win.

---

In short, I think the best approach would be to merge this PR (if you are convinced the answers are right) and then consider use of `WindowExec` for null-as-missing linear interpolation (#7287) and maybe for filling with previous as well. Does that make sense?

> The upside to using the WindowExec approach is that there was quite a lot of effort put into them to make them efficient and parallelized (partitioned) and streaming

Currently `GapFillExec` requires input to be single-partitioned, but I think it could be updated to be able to run in parallel provided that the groups (not including `time` column) are on the same partition. I wonder if that would be hard to do. It seems like _maybe_ it's just a matter of implementing `ExecutionPlan::required_input_distribution` in the right way?

**Comment by alamb** [comment]
> The user put the null values in the input on purpose, and doesn't want them filled

The "null as intentional" vs "nulls as missing" makes total sense to me was a subtlety that I did not appreciate. I think the idea originally in SQL was that NULL was meant to model missing values, but I can see how people might want to distinguish between NULLs added by gapfilling and ones that were originally in their data. 

> In short, I think the best approach would be to merge this PR (if you are convinced the answers are right) and then consider use of WindowExec for null-as-missing linear interpolation (https://github.com/influxdata/influxdb_iox/issues/7287) and maybe for filling with previous as well. Does that make sense?

Yes that makes perfect sense

> Currently GapFillExec requires input to be single-partitioned, but I think it could be updated to be able to run in parallel provided that the groups (not including time column) are on the same partition. I wonder if that would be hard to do. It seems like maybe it's just a matter of implementing ExecutionPlan::required_input_distribution in the right way?

I agree

**Review by alamb — APPROVED** [review]
🚢 🇮🇹

File: iox_query/src/exec/gapfill/mod.rs (hunk 8 above)
Comment (wolffcm) [reply iox_query/src/exec/gapfill/mod.rs:79]: Sadly it is, as these strategies are not used outside of tests yet, that will change with #7071.
